libraries/AC_PID/AC_PID_2D.cpp

```diff
@@ -55,158 +60,139 @@ AC_PID_2D::AC_PID_2D(float initial_p, float initial_i, float initial_d, float in
     _kp = initial_p;
     _ki = initial_i;
     _kd = initial_d;
-    _imax = fabsf(initial_imax);
-    filt_hz(initial_filt_hz);
-    filt_d_hz(initial_filt_d_hz);
+    _kff = initial_ff;
+    _kimax = fabsf(initial_imax);
+    filt_E_hz(initial_filt_E_hz);
+    filt_D_hz(initial_filt_D_hz);
 
-    // reset input filter to first value received and derivitive to zero
-    reset_filter();
+    // reset input filter to first value received
+    _flags._reset_filter = true;
 }
 
 // set_dt - set time step in seconds
 void AC_PID_2D::set_dt(float dt)
 {
     // set dt and calculate the input filter alpha
     _dt = dt;
-    calc_filt_alpha();
-    calc_filt_alpha_d();
 }
 
-// filt_hz - set input filter hz
-void AC_PID_2D::filt_hz(float hz)
+// filt_E_hz - set error filter hz
+void AC_PID_2D::filt_E_hz(float hz)
 {
-    _filt_hz.set(fabsf(hz));
-
-    // sanity check _filt_hz
-    _filt_hz = MAX(_filt_hz, AC_PID_2D_FILT_HZ_MIN);
-
-    // calculate the input filter alpha
-    calc_filt_alpha();
+    _filt_E_hz.set(fabsf(hz));
 }
 
-// filt_d_hz - set input filter hz
-void AC_PID_2D::filt_d_hz(float hz)
+// filt_D_hz - set derivative filter hz
+void AC_PID_2D::filt_D_hz(float hz)
 {
-    _filt_d_hz.set(fabsf(hz));
-
-    // sanity check _filt_hz
-    _filt_d_hz = MAX(_filt_d_hz, AC_PID_2D_FILT_D_HZ_MIN);
-
-    // calculate the input filter alpha
-    calc_filt_alpha_d();
+    _filt_D_hz.set(fabsf(hz));
 }
 
-// set_input - set input to PID controller
-//  input is filtered before the PID controllers are run
-//  this should be called before any other calls to get_p, get_i or get_d
-void AC_PID_2D::set_input(const Vector2f &input)
+//  update_all - set target and measured inputs to PID controller and calculate outputs
+//  target and error are filtered
+//  the derivative is then calculated and filtered
+//  the integral is then updated based on the setting of the limit flag
+Vector2f AC_PID_2D::update_all(float target_x, float target_y, Vector2f measurement, bool limit)
 {
+    _target = Vector2f(target_x, target_y);
     // don't process inf or NaN
-    if (!isfinite(input.x) || !isfinite(input.y)) {
-        return;
+    if (!isfinite(_target.x) || !isfinite(_target.y) || !isfinite(measurement.x) || !isfinite(measurement.y)) {
```

Comment:
We also do NaN checks in the 1D equivalent.

Reply:
Can a float be both nan and finite?  In any case, I've added isnan checks and also moved the setting of _target to after these checks.  Doesn't seem right that we would set the target *before* checking the target values are reasonable.